File: 03-march/14-the-k-th-lexicographical-string-of-all-happy-strings-of-length-n/solution.js
/**
 * @param {number} n
 * @param {number} k
 * @return {string}
 */
var getHappyString = function(n, k) {
  const total = 3 * (1 << (n - 1));
  if (k > total) {
    return "";
  }

  const chars = [];
  let prev = "";
  k--;

  for (let i = 0; i < n; i++) {
    const choices = [];
    for (const ch of "abc") {
      if (ch !== prev) {
        choices.push(ch);
      }
    }

    const block = 1 << (n - i - 1);
    for (const ch of choices) {
      if (k >= block) {
        k -= block;
      } else {
        chars.push(ch);
        prev = ch;
        break;
      }
    }
  }

  return chars.join("");
};

File: 03-march/14-the-k-th-lexicographical-string-of-all-happy-strings-of-length-n/solution.py
class Solution:
    def getHappyString(self, n: int, k: int) -> str:
        total = 3 << (n - 1)
        if k > total:
            return ""

        chars = []
        prev = ""
        k -= 1

        for i in range(n):
            choices = [c for c in "abc" if c != prev]
            block = 1 << (n - i - 1)
            for ch in choices:
                if k >= block:
                    k -= block
                else:
                    chars.append(ch)
                    prev = ch
                    break

        return "".join(chars)
